internal/db/schema/migrations/oss/postgres/70/02_sessions.up.sql

```diff
@@ -0,0 +1,154 @@
+begin;
+
+  -- Replaces trigger from 56/02_add_data_key_foreign_key_references.up.sql
+  drop trigger cancel_session_with_null_fk on session;
+  create or replace function cancel_session_with_null_fk() returns trigger
+  as $$
+  begin
+    -- Note that we need each of these to run in case
+    -- more than one of them is null.
+    if new.auth_token_id is null then
+      perform cancel_session(new.public_id);
+    end if;
+    if new.project_id is null then
+      -- Setting the key_id to null will allow the scope
+      -- to cascade delete its keys.
+      new.key_id = null;
+      perform cancel_session(new.public_id);
+    end if;
+    if new.target_id is null then
+      perform cancel_session(new.public_id);
+    end if;
+    if new.user_id is null then
+      perform cancel_session(new.public_id);
+    end if;
+    return new;
+  end;
+  $$ language plpgsql;
+  
+  create trigger cancel_session_with_null_fk before update of auth_token_id, project_id, target_id, user_id on session
+    for each row execute procedure cancel_session_with_null_fk();
+
+  create table session_target_address (
+    session_id wt_public_id,
+    target_id wt_public_id,
+    constraint target_address_fkey foreign key (target_id)
+        references target_address (target_id)
+        on delete set null
+        on update cascade,
+    constraint session_fkey foreign key (session_id)
+        references session (public_id)
+        on delete cascade
+        on update cascade,
+    constraint session_id_target_id_uq
+        unique(session_id, target_id)
+  );
+  comment on table session_target_address is
+    'session_target_address entries represent a session that is using a network address that is assigned directly to a Target.';
+
+  create trigger immutable_columns before update on session_target_address
+    for each row execute procedure immutable_columns('session_id');
+
+  create or replace function cancel_session_with_null_target_address_fk() returns trigger
+  as $$
+  begin
+    if new.target_id is null then
+      perform cancel_session(new.session_id);
+      delete from session_target_address where session_id = new.session_id;
+    end if;
+    return new;
+  end;
+  $$ language plpgsql;
+  
+  create trigger cancel_session_with_null_target_address_fk after update of target_id on session_target_address
+    for each row execute procedure cancel_session_with_null_target_address_fk();
```

Comment:
I can set the foreign key reference to `on delete cascade` for both session_target_address & session_host_set_host, then I would just need to update the trigger so that it gets invoked after a delete event. At that point we would only need a single function for both tables. I can change the function name to cancel_session_on_deleted_resource(). An example of this:
```
create or replace function cancel_session_with_deleted_resource() returns trigger
  as $$
  begin
    perform cancel_session(old.session_id);
    return old;
  end;
  $$ language plpgsql;
  
  create trigger cancel_session_with_deleted_target_address_fk after delete on session_target_address
    for each row execute procedure cancel_session_with_deleted_resource();
```

My only problem with the example posted above is that it would be invoked whenever the row is deleted, regardless of why the row was deleted. With the current implementation in this PR, canceling the session is enforced to only when the foreign key reference is removed by deleting the resource itself.